### PR TITLE
Remove excessive selection updating for SongSelect

### LINKED ISSUE #4818: Can't change beatmap while importing beatmaps from stable
**Describe the bug:** Cann't change beatmap while importing beatmaps from stable

**Screenshots or videos showing encountered issue:** https://streamable.com/e9jbi

**osu!lazer version:** 2019.514.0

**Logs:** N/A

### PR DESCRIPTION
**Pull request by nyquillerium**
Fixes https://github.com/ppy/osu/issues/4818

Previously, the selected beatmap would be updated every time an import occurred. While this ensured that the selected beatmap would always be what was currently playing, it made it impossible to select beatmaps while a large amount of beatmaps are imported.

Moving the check for whether or not a beatmap has already been selected to before this happens resolves this. 

This logic was used by three scenarios:
 * Hiding/Restoring beatmaps.
 * Importing a new beatmap.
 * Entering songselect

I've tested this logic with both empty/full playlists using the above scenarios, and they seem to work as before.

### CODE REVIEW

**Comment by peppy** [comment]
Broke current selection after hiding a difficulty (first difficulty will be selected, changing the user's selection).

File: osu.Game/Screens/Select/BeatmapCarousel.cs [hunk 1]
```diff
@@ -158,6 +158,9 @@ public void UpdateBeatmapSet(BeatmapSetInfo beatmapSet)
 
                 var newSet = createCarouselSet(beatmapSet);
 
+                // Since we're about to remove the selected beatmap, store its ID so we can go back if needed.
+                var previouslySelectedID = selectedBeatmap?.Beatmap.ID;
```
Comment (peppy) [inline osu.Game/Screens/Select/BeatmapCarousel.cs]: couldn't this replace `hadSelection`?

File: osu.Game.Tests/Visual/SongSelect/TestScenePlaySongSelect.cs [hunk 2]
```diff
@@ -209,7 +209,45 @@ public void TestStartAfterUnMatchingFilterDoesNotStart()
...
+            int? previousID = null;
+            createSongSelect();
+            importForRuleset(0);
+            AddStep("Move to last difficulty", () => songSelect.Carousel.SelectBeatmap(songSelect.Carousel.BeatmapSets.First().Beatmaps.Last()));
+            AddStep("Store current ID", () => previousID = songSelect.Carousel.SelectedBeatmap.ID);
+            AddStep("Hide first beatmap", () => manager.Hide(songSelect.Carousel.SelectedBeatmapSet.Beatmaps.First()));
+            AddAssert("Selected beatmap has not changed", () => songSelect.Carousel.SelectedBeatmap.ID == previousID);
+        }
+
+        private void importForRuleset(int id) => AddStep($"import test map for ruleset {id}",
```
Comment (peppy) [inline osu.Game.Tests/Visual/SongSelect/TestScenePlaySongSelect.cs]: This is duplicate code. Why was the original call site not replaced with a call to this?

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Your test pasts on master. It is not testing the issue at hand correctly.

**Comment by nyquillerium** [comment]
Which one? The second test passes on master because the issue isn't visible functionality-wise there, which is what it is testing. I would have to expose deeper values to test otherwise (such as the local `previouslySelectedID`). The first test definitely did not pass on master when I checked, I'll check again.

**Comment by nyquillerium** [comment]
Ah, the first test got changed into something that was wrong somewhere along the way. I tested again and this one works.

File: .idea/.idea.osu/.idea/.name [hunk 3]
```diff
@@ -0,0 +1 @@
+osu
```
Comment (peppy) [inline .idea/.idea.osu/.idea/.name]: ?

File: osu.Game.Tests/Visual/SongSelect/TestScenePlaySongSelect.cs [hunk 4]
```diff
@@ -215,11 +216,13 @@ public void TestAddNewBeatmapWhileSelectingRandom()
...
             int beatmapChangedCount = 0;
+            int debounceCount = 0;
             createSongSelect();
-            AddStep("Setup counter", () =>
+            AddStep("Setup counters", () =>
             {
                 beatmapChangedCount = 0;
-                songSelect.Carousel.SelectionChanged += _ => beatmapChangedCount++;
+                debounceCount = 0;
+                songSelect.Carousel.SelectionChanged += _ => beatmapChangedCount += 1;
```
Comment (peppy) [inline osu.Game.Tests/Visual/SongSelect/TestScenePlaySongSelect.cs]: what was the reason for changing this?

**Review by smoogipoo — APPROVED** [review]
Seems ok to me

**Pull request merged**